sprint_12

Vlad-Gorelov opened the pull request.

Commit:
- Add decoding models for JSON + convert model for UI
- Created new class ImagesListService with network layer to upload photos from unsplash

Commit:
Loading images to cell in table

Commit:
Bug fixed + small refactoring

Commit:
Bug fixed

Commit:
Likes button function

Vlad-Gorelov committed:
Fixed single image screen

Vlad-Gorelov committed:
Fixed bug

Vlad-Gorelov committed:
Bug fixed

Vlad-Gorelov committed:
Add microanimations on like button

DanilaG commented on PhotoFeed/ImagesList/PhotoResult.swift:
```diff
@@ -0,0 +1,46 @@
+    let size: CGSize
+    let createdAt: Date?
+    let description: String?
+    let thumbImageURL: String?
+    let fullImageURL: String?
```
Давай в модельке `Photo` url будем хранить в виде `URL?`. Так как дальше, при использовании в UI и логике так или иначе всегда будем переводить в URL.

Vlad-Gorelov replied:
не понял комментария

DanilaG replied:
Тут я предлогаяю вместо:
```
struct Photo {
    let id: String
    let size: CGSize
    let createdAt: Date?
    let description: String?
    let thumbImageURL: String?
    let fullImageURL: String?
    let isLiked: Bool
}
```
сделать:
```
struct Photo {
    let id: String
    let size: CGSize
    let createdAt: Date?
    let description: String?
    let thumbImageURL: URL?
    let fullImageURL: URL?
    let isLiked: Bool
}
```
Сейчас URL-ы перед непосредственным моментом использования переводятся из `String` в `URL` (например в `configCell`). Но так как на тех уровнях абстрацкии, где мы работаем с `Photo` нам различные URL нужне не в виде `String`, а в виде `URL`, то предлагаю в `Photo` для хранения url использовать тип данных `URL`, а не `String`

DanilaG commented on PhotoFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,128 @@
+    private func decodedResult(_ photoResult: PhotoResult) -> Photo {
+        return Photo.init(id: photoResult.id,
+                          size: CGSize(width: photoResult.width ?? 0, height: photoResult.height ?? 0),
+                          createdAt: dateFormater.date(from: photoResult.createdAt ?? ""),
```
Можно заменить на такую конструкцию, тогда не придётся вызывать форматер с пустой строкой:
```
photoResult.createdAt.flatMap({ dateFormater.date(from: $0) })
```

Vlad-Gorelov replied:
что такое flatMap и откуда оно берется?

DanilaG replied:
`flatMap` - это метод опционалов. Он позволяет "мепить" (переводить) значения опционалов, а если значение отсутсвует то возвращает nil. 

В нашем случае `createdAt` - опционал, у него можем вызвать метод `flatMap`. Если `createdAt` == nil - то `flatMap` вернёт `nil`, если `createdAt` не nil, то для значения опционала выполнится замыкание переданное во `flatMap` (в нашем случае для строки с датой выполнится перевод в дату при помощи форматера) и `flatMap` вернёт результат исполнения замыкания (в нашем случае вернёт результат форматирования - `Date?`).  

Пример для понимания:
```
var a: Int?
...
let newValue = a.flatMap({ _ in 1 })
```
В данном случае `newValue` будет nil если `a` == nil, и если `a` != nil то всегда будет 1 (не зависит от значения)

Подробнее про метод можно прочитать [тут](https://developer.apple.com/documentation/swift/optional/flatmap(_:))

DanilaG commented on PhotoFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,128 @@
+                          createdAt: dateFormater.date(from: photoResult.createdAt ?? ""),
+                          description: photoResult.description,
+                          thumbImageURL: photoResult.urls?.thumbImageURL,
+                          fullImageURL: photoResult.urls?.fullImageURL,
```
Давай вот тут будем сразу переводить в URL

DanilaG commented on PhotoFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,128 @@
+
+final class ImagesListService {
+
+    static let didChangeNotification = Notification.Name(rawValue: "ImagesListServiceDidChange")
```
🔥, что назвал с маленькой буквы)

DanilaG commented on PhotoFeed/Helpers/Constans.swift:
```diff
@@ -0,0 +1,12 @@
+import UIKit
+
+enum Constants {
+    static let accessKey = "tzdwsi8S2ctMuNvS0SdMqR61I_jmvKY-s3yDA4-UJ6c"
+    static let secretKey = "7cL8yUdneOy1d6czQC29oHou726mhsa41wn9MgozXNA"
+    
+    static let redirectURI = "urn:ietf:wg:oauth:2.0:oob"
+    static let accessScope = "public+read_user+write_likes"
+    static let defaultBaseUrl: String = "https://api.unsplash.com"
+    static let accessTokenURL = "https://unsplash.com/oauth/token"
+    static let unsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
+}
```
🔥, что все ключи в константах и в отдельном name space. Это упрощает работу с ними и снижает вероятность на опечатку)

DanilaG commented on PhotoFeed/Helpers/DateFormatter.swift:
```diff
@@ -0,0 +1,8 @@
+import Foundation
+
+let dateFormatter: DateFormatter = {
+    let dateFormatter = DateFormatter()
+    dateFormatter.dateStyle = .long
+    dateFormatter.timeStyle = .none
+    return dateFormatter
+}()
```
Давай также создадим name space для форматеров по умолчанию?)

Vlad-Gorelov replied:
я не понимаю комментарий

DanilaG replied:
Предлагаю по аналагии с тем как сделано для различных констант завести [namespace](https://ru.wikipedia.org/wiki/Пространство_имён_(программирование)) для форматеров дат.
Например, он может выглядить так:
```
enum DateFormatters {
    
    static let dayMonthYear: DateFormatter = {
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .long
        dateFormatter.timeStyle = .none
        return dateFormatter
    }()
    
    static let iso = ISO8601DateFormatter()
}
```
Тогда обращение к ним будет выглядить более красиво + сохраним условие, что все форматеры создаются единожды:
`DateFormatters.dayMonthYear.string(from: date)`

Vlad-Gorelov committed:
Исправлены критические замечания:

-Раньше, если дважды вызвать fetchPhotoNextPage до завершения загрузки, то первая загрузка отменялась и стартовала новая для этой же страницы. Теперь мы не отменяем старую загрузку, а просто не запускаем новую (guard task == nil else { return }), так мы получим ответ быстрее, так как старый запрос начался раньше)

-Поправил текст алерта при выходе из профиля в соответствии с макетом

Дополнительные комментарии: изменён цвет на лаунчскрине на ypBlack

Vlad-Gorelov committed:
Bug fixed
